src/DataStructures/DataBox/DataBox.hpp

```diff
@@ -1621,102 +1364,42 @@ constexpr bool has_return_tags_and_argument_tags_v =
-                             DataBox_detail::check_mutate_apply_argument_tags(
-                                 BoxTags{}, ArgumentTags{}) and
-                             noexcept(DataBox_detail::mutate_apply(
-                                 f, box, MutateTags{}, ArgumentTags{},
-                                 std::forward<Args>(args)...))) {
-  // These checks are duplicated in the noexcept specification above
-  // because the noexcept(DataBox_detail::mutate_apply(...)) can cause
-  // a compilation error before the checks in the function body are
-  // performed.
-  DataBox_detail::check_mutate_apply_mutate_tags(BoxTags{}, MutateTags{});
-  DataBox_detail::check_mutate_apply_argument_tags(BoxTags{}, ArgumentTags{});
-  return DataBox_detail::mutate_apply(std::forward<F>(f), box, MutateTags{},
-                                      ArgumentTags{},
-                                      std::forward<Args>(args)...);
+    Args&&... args) noexcept {
+  detail::check_tags_are_in_databox(BoxTags{}, MutateTags{});
+  detail::check_tags_are_in_databox(BoxTags{}, ArgumentTags{});
+  detail::mutate_apply(std::forward<F>(f), box, MutateTags{}, ArgumentTags{},
+                       std::forward<Args>(args)...);
 }
 
-template <typename F, typename BoxTags, typename... Args,
-          Requires<DataBox_detail::has_return_tags_and_argument_tags_v<
-              std::decay_t<F>>> = nullptr,
-          typename MutateTags = typename std::decay_t<F>::return_tags,
-          typename ArgumentTags = typename std::decay_t<F>::argument_tags>
-SPECTRE_ALWAYS_INLINE constexpr auto mutate_apply(
+template <typename F, typename BoxTags, typename... Args>
+SPECTRE_ALWAYS_INLINE constexpr void mutate_apply(
     F&& f, const gsl::not_null<DataBox<BoxTags>*> box,
-    Args&&... args) noexcept(DataBox_detail::
-                                 check_mutate_apply_mutate_tags(
-                                     BoxTags{}, MutateTags{}) and
-                             DataBox_detail::check_mutate_apply_argument_tags(
-                                 BoxTags{}, ArgumentTags{}) and
-                             noexcept(DataBox_detail::mutate_apply(
-                                 f, box, MutateTags{}, ArgumentTags{},
-                                 std::forward<Args>(args)...))) {
-  // These checks are duplicated in the noexcept specification above
-  // because the noexcept(DataBox_detail::mutate_apply(...)) can cause
-  // a compilation error before the checks in the function body are
-  // performed.
-  DataBox_detail::check_mutate_apply_mutate_tags(BoxTags{}, MutateTags{});
-  DataBox_detail::check_mutate_apply_argument_tags(BoxTags{}, ArgumentTags{});
-  return DataBox_detail::mutate_apply(std::forward<F>(f), box, MutateTags{},
-                                      ArgumentTags{},
-                                      std::forward<Args>(args)...);
+    Args&&... args) noexcept {
+  mutate_apply<typename std::decay_t<F>::return_tags,
```

Comment:
If I understand correctly you're removing the possibility that `mutate_apply` returns values. We're probably not using that functionality anywhere right now, but it could be useful at some point, don't you think?

Reply:
no, detail::mutate_apply doesn't return anything (it just calls db::mutate)